From 4c69a421d4874e5f3c8ffde5e44cab33b44a608e Mon Sep 17 00:00:00 2001
From: Christopher Caldwell <chrisolof@gmail.com>
Date: Tue, 17 Jul 2018 12:04:49 -0600
Subject: [PATCH] Open fieldgroups that contain elements with errors

---
 field_group.module | 50 +++++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 45 insertions(+), 5 deletions(-)

diff --git a/field_group.module b/field_group.module
index 612c3e7..88fa0f6 100644
--- a/field_group.module
+++ b/field_group.module
@@ -191,6 +191,7 @@ function field_group_form_alter(array &$form, FormStateInterface $form_state) {
 
       field_group_attach_groups($form, $context);
       $form['#process'][] = 'field_group_form_process';
+      $form['#pre_render'][] = 'field_group_form_pre_render';
     }
   }
 
@@ -249,15 +250,54 @@ function field_group_entity_view_alter(&$build, EntityInterface $entity, EntityD
 }
 
 /**
- * Pre render callback for rendering groups.
+ * Pre render callback for rendering groups in forms.
  *
  * @param array $element
- *   Form that is being rendered.
- *
- * @deprecated Use field_group_form_process instead.
+ *   The form element that is being rendered.
  */
 function field_group_form_pre_render(array $element) {
-  return field_group_form_process($element);
+  // Open any closed field groups that contain elements with errors.
+  if (!empty($element['#fieldgroups'])) {
+    foreach ($element['#fieldgroups'] as $fieldgroup) {
+      $closed = isset($element[$fieldgroup->group_name]['#open']) && !$element[$fieldgroup->group_name]['#open'];
+      if ($closed) {
+        foreach ($fieldgroup->children as $child) {
+          if (field_group_elements_contain_errors($element[$child])) {
+            $element[$fieldgroup->group_name]['#open'] = TRUE;
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  return $element;
+}
+
+/**
+ * Determines if an elements array contains validation errors.
+ *
+ * @param array $elements
+ *   The elements array to check for errors.
+ *
+ * @return bool
+ *   TRUE if the elements array contains validation errors, otherwise FALSE.
+ */
+function field_group_elements_contain_errors(array $elements) {
+  // Any errors at this level of the elements array?
+  if (!empty($elements['#errors']) || !empty($elements['#children_errors'])) {
+    return TRUE;
+  }
+
+  // Dive down.
+  foreach (Element::children($elements) as $child) {
+    if (field_group_elements_contain_errors($elements[$child])) {
+      return TRUE;
+    }
+  }
+
+  // No errors.
+  return FALSE;
 }
 
 /**
-- 
2.17.0
