diff --git a/cyberwoven_alert_bar.module b/cyberwoven_alert_bar.module
index 26c3866..6815866 100644
--- a/cyberwoven_alert_bar.module
+++ b/cyberwoven_alert_bar.module
@@ -42,6 +42,7 @@ function cyberwoven_alert_bar_theme($existing, $type, $theme, $path) {
         'more_link_external' => null,
         'alert_color_scheme' => null,
         'additional_classes' => null,
+        'unique_id' => null,
       ],
     ],
   ];
@@ -70,6 +71,7 @@ function cyberwoven_alert_bar_preprocess_block__cyberwoven_alert_bar(&$variables
     'more_link_external' => $config->get('more_link_external'),
     'alert_color_scheme' => $config->get('alert_color_scheme'),
     'additional_classes' => $config->get('additional_classes'),
+    'unique_id' => $config->get('unique_id'),
   ];
 
 }
diff --git a/js/cyberwoven_alert_bar.js b/js/cyberwoven_alert_bar.js
index 0798a8e..3b77dd8 100644
--- a/js/cyberwoven_alert_bar.js
+++ b/js/cyberwoven_alert_bar.js
@@ -1,8 +1,8 @@
 (function ($, Drupal) {
   Drupal.behaviors.cyberwovenAlertBar = {
     attach: function (context, settings) {
-
-      var cyberwoven_alert_bar_closed = $.cookie('cyberwoven_alert_bar_closed', Number);
+      var cookieintro = 'cyberwoven_alert_bar_closed_';
+      var cyberwoven_alert_bar_closed = $.cookie(cookieintro.concat(settings.cw_alert_bar.unique_id), Number);
       if (!cyberwoven_alert_bar_closed) {
         var current_time = Math.round(new Date().getTime()/1000);
         if (current_time < drupalSettings.alertSettings.expires) {
@@ -14,7 +14,7 @@
         e.preventDefault();
         if ($('#cw-alert-bar').length > 0) {
           $('#cw-alert-bar').removeClass('show-alert');
-          $.cookie('cyberwoven_alert_bar_closed', 1);
+          $.cookie(cookieintro.concat(settings.cw_alert_bar.unique_id), 1);
         }
       });
 
diff --git a/src/Form/CyberwovenAlertBarSettingsForm.php b/src/Form/CyberwovenAlertBarSettingsForm.php
index 3863f3e..b6db840 100644
--- a/src/Form/CyberwovenAlertBarSettingsForm.php
+++ b/src/Form/CyberwovenAlertBarSettingsForm.php
@@ -157,6 +157,7 @@ class CyberwovenAlertBarSettingsForm extends ConfigFormBase {
         ->set('more_link_external', $form_state->getValue('cyberwoven_alert_bar_more_link_external'))
         ->set('alert_color_scheme', $form_state->getValue('cyberwoven_alert_bar_alert_color_scheme'))
         ->set('additional_classes', $form_state->getValue('cyberwoven_alert_bar_additional_classes'))
+        ->set('unique_id', uniqid())
         ->save();
 
       Cache::invalidateTags(array('config:cyberwoven_alert_bar.settings'));
diff --git a/src/Plugin/Block/CyberwovenAlertBarBlock.php b/src/Plugin/Block/CyberwovenAlertBarBlock.php
index c9b16bd..a671219 100644
--- a/src/Plugin/Block/CyberwovenAlertBarBlock.php
+++ b/src/Plugin/Block/CyberwovenAlertBarBlock.php
@@ -45,6 +45,7 @@ class CyberwovenAlertBarBlock extends BlockBase implements BlockPluginInterface
       '#more_link_external' => $config->get('more_link_external'),
       '#alert_color_scheme' => $config->get('alert_color_scheme'),
       '#additional_classes' => $config->get('additional_classes'),
+      '#unique_id' => $config->get('unique_id'),
       '#cache' => [
         'tags' => ['config:cyberwoven_alert_bar.settings'],
       ]
@@ -56,6 +57,7 @@ class CyberwovenAlertBarBlock extends BlockBase implements BlockPluginInterface
       $build['cw-alert-bar'] = null;
     }
 
+    $build['#attached']['drupalSettings']['cw_alert_bar']['unique_id']= $config->get('unique_id');
     return $build;
 
   }
